src/quicklist.c

```diff
@@ -1171,7 +1147,7 @@ int quicklistDelRange(quicklist *quicklist, const long start,
             quicklist->count -= del;
             quicklistDeleteIfEmpty(quicklist, node);
             if (node)
-                quicklistRecompressOnly(node);
+                quicklistCompress(quicklist, node);
```

Comment:
Why are you using `quicklistCompress` here and in other places? Because `node->recompress` is 0 and can't compress?  
This change may hide other bugs where node is not compressed.

Reply:
This is mentioned in commit log, This change can fix #12563,  for more efficient experience, quicklistNode->recompress should be redesigned.

Reply:
Do you mean that `quicklistDelRange` also can cause uncompress nodes?
Could you provide the smoke test and others?

Reply:
No, this means I can't say that quicklistDelRange() won't cause uncompressed nodes. If anyone can prove that, should leave a comment here and keep using quicklistRecompressOnly().

Reply:
Note that the uncompressed node is only created because we reset `iter->node` early before `quicklistReleaseIterator()`, not because of the uncompression caused by `quicklistRecompressOnly`.
I'd prefer to fix the problem of creating uncompressed nodes, rather than using `quicklistRecompressOnly` for all potentially uncompressed nodes.

Reply:
It is caused by quicklistRecompressOnly(). It is caused by 'quicklistNode->recompress' can't tell a node is in the compress depth or not, it can only tell a node is compressed before use or not. Because a node can remain uncompressed even within compression depth. So I sad we should redesign 'quicklistNode->recompress'.

Reply:
Please note the code in `quicklistReleaseIterator() -> quicklistCompress()`
```
        if ((_node)->recompress)                                               \
            quicklistCompressNode((_node));                                    \
        else                                                                   \
            __quicklistCompress((_ql), (_node));                               \
```
When releasing the quicklist iterator, if `node->recompress` is 0, we will upgrade using `__quicklistCompress()`.

Reply:
Yes if we use a node temporarily, it's ok to use quicklistRecompressOnly() for recompress, but we changed node's entry here.

Reply:
There will be the following 2 scenarios here
1) node is uncompressed
    After it is deleted part of the data, it will not be compressed.
2) the node is compressed
   - Compression condition is not satisfied after deleting elements
   - QuicklistRecompressOnly() can still compress the node after deleting the element and succeeds.

I don't see any problem to use `quicklistCompress`.
Am i missing something?

Reply:
The first scenario, I don't know if it's possible that a uncompressed node can be compressed after deleted part of the data.

Reply:
if so we should find out why this node wasn't compressed correctly.

Reply:
This node is compressed correctly, but it is not compress small enough, so it remains uncompressed. What I don't know is if it's possible to compress it small enough after deleted part of the data.

Reply:
Compression algorithms are usually unlikely to do this, when a piece of data has a low compression ratio, it means that there are too few duplicate blocks, and when some of the blocks are removed, its compression ratio should become lower.

Reply:
OK, Then I will roll back this change.